keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2043 @@
+node is found.
+
+The hardware that is expected to need this more flexible allocation
+approach is going to be used by only a small subset of the pods in the
+cluster and those pods are likely to run for extended periods of time,
+so this is not a major concern.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Implementation
+
+![components](./components.png)
+
+Several components must be implemented or modified in Kubernetes:
+- The new API must be added to kube-apiserver.
+- A new controller in kube-controller-manager which creates 
+  ResourceClaims from Pod ResourceClaimTemplates, similar to
+  https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/volume/ephemeral
+- A kube-scheduler plugin must detect Pods which reference a
+  ResourceClaim (directly or through a template) and ensure that the
+  resource is allocated before the Pod gets scheduled, similar to
+  https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/scheduling/scheduler_binder.go
+- Kubelet must be extended to retrieve information from ResourceClaims
+  and to call a resource kubelet plugin. That plugin returns CDI device ID(s)
+  which then must be passed to the container runtime.
+
+For a resource driver the following components are needed:
+- *Resource driver controller*: a central component which handles resource allocation
+  by watching ResourceClaims and updating their status once it is done with
+  allocation.
+- *Resource kubelet plugin*: a component which cooperates with kubelet to prepare
+  the usage of the resource on a node.
+
+The utility library will be developed outside of Kubernetes and does not have
+to be used by drivers, therefore it is not described further in this KEP.
+
+### Resource allocation flow
+
+The following diagram shows how resource allocation works for a resource that
+gets defined inline in a Pod. For a full definition of ResourceClass,
+ResourceClaim and ResourceClaimTemplate see the [API](#API) section below.
+
+Several of these operations may fail without changing the system state. They
+then must be retried until they succeed or something else changes in the
+system, like for example deleting objects. These additional state transitions
+are not shown for the sake of simplicity.
+
+![allocation](./allocation.png)
```

Comment:
It's a very nice picture, but I can't easily comment on it or search easily.  We don't have a lot of examples of KEPs which use images, for this reason.  Looking at the image I feel like it could be just as clear as pseudocode.

Reply:
The source for the image is part of the PR, see [allocation.puml](https://github.com/kubernetes/enhancements/pull/3064/files#diff-4d0bd211a43641ac2b4164a2e40ec08490b0d0860d8331047c89a3987eceb74b). It's plain text, with rules in the Makefile to build this picture. It uses the PlantUML web service, so no local tools are needed besides Go to build the client for that service.

Is that good enough (can be commented on, can be modified easily later on) or should I replace with pseudo code?

Reply:
The current revision uses a flow with if/else checks instead of the the diagram.